composer/datasets/dataloader.py

```diff
@@ -79,47 +78,69 @@ def __next__(self) -> Batch:
+            2 means there will be a total of 2 * num_workers samples prefetched across all workers.
+        persistent_workers (bool): Whether or not to shutdown workers after the dataset has been consumed once.
+        pin_memory (bool): Whether or not to copy Tensors into CUDA pinned memory before returning them.
+        timeout (float): Timeout, in seconds, for collecting a batch from workers. Set to 0 for no timeout.
+    
+    """
 
-    num_workers: int = hp.required(doc="Number of CPU workers to use per gpu", template_default=8)
-    prefetch_factor: int = hp.required(doc="Number of samples loaded in advance by each worker", template_default=2)
-    persistent_workers: bool = hp.required(doc="Whether or not to shutdown workers after the dataset"
-                                           " has been consumed once",
+    num_workers: int = hp.required("Number of CPU workers to use per device to fetch data.", template_default=8)
+    prefetch_factor: int = hp.required("Number of samples loaded in advance by each worker", template_default=2)
+    persistent_workers: bool = hp.required(textwrap.dedent("""Whether or not to shutdown workers after the dataset
+        has been consumed once"""),
                                            template_default=True)
-    pin_memory: bool = hp.required(doc="Whether or not to copy Tensors into CUDA pinned memory"
-                                   " before returning them",
+    pin_memory: bool = hp.required(textwrap.dedent("""Whether or not to copy Tensors into CUDA pinned memory
+        before returning them"""),
                                    template_default=True)
-    timeout: int = hp.required(doc="Timeout value for collecting a batch from workers. 0 for no timeout.",
-                               template_default=0)
+    timeout: float = hp.required("Timeout, in seconds, for collecting a batch from workers. Set to 0 for no timeout",
+                                 template_default=0)
 
     def initialize_object(
         self,
+        dataset: Dataset,
+        *,
         batch_size: int,
-        sampler: Sampler,
-        dataloader_spec: DataloaderSpec,
+        sampler: torch.utils.data.Sampler[int],
+        drop_last: bool,
+        collate_fn: Optional[Callable] = None,
+        worker_init_fn: Optional[Callable] = None,
     ) -> DataLoader:
-        """Initializes the dataloader."""
-
-        dataloader = torch.utils.data.DataLoader(
-            dataloader_spec.dataset,
-            batch_size=batch_size,
-            shuffle=False,  # set in the sampler
-            num_workers=self.num_workers,
-            pin_memory=self.pin_memory,
-            drop_last=dataloader_spec.drop_last,
-            sampler=sampler,
-            collate_fn=dataloader_spec.collate_fn,
-            worker_init_fn=dataloader_spec.worker_init_fn,
-            multiprocessing_context=dataloader_spec.multiprocessing_context,
-            generator=dataloader_spec.generator,
```

Comment:
> This looks good - I'm very happy to see `DataloaderSpec` losing favor. The only thing I wonder is whether `split_fn` and `device_transform_fn` could be removed. The former I think is unnecessary if we just load `N` microbatches instead of 1 batch, and the latter could be replaced with explicit augmentations?

Re: split_fn, I think we want to have the dataloader yield entire batches instead of microbatches, but @abhi-mosaic can speak more to this one.

Re: device_transform_fn: The transforms that happen in the dataset (e.g. resnet crop/resize) happen on the cpu. This allows for GPU transformations that are specified by the dataset (like what imagenet does with the batchnorm).

Reply:
Multiprocessing context is undocumented so I presume it's deprecated? https://pytorch.org/docs/stable/data.html

generator -- Since we're passing a sampler to the dataloader, the generator argument is unused. The dataset can specify a custom generator in the sampler constructor, if desired.

Reply:
I think it may be simpler to keep our system of loading full batches for now... but probably worth revisiting later. Now that we're getting into multi-node training, if we run a workload on 1 vs. 8 vs. 64 GPUs, the visible per_gpu_batch_size will be different for any algorithm like Mixup that would like to perform an operation across-batch. So should we make it possible to always alter the full batch (across GPUs), or should we expose the microbatching somehow to algorithms....